src/vmm/src/lib.rs

```diff
@@ -0,0 +1,404 @@
+
+        // Set up the speaker PIT, because some kernels are musical and access the speaker port
+        // during boot. Without this, KVM would continuously exit to userspace.
+        // TODO: does this *need* the irqchip in place?
+        let mut pit_config = kvm_pit_config::default();
+        pit_config.flags = KVM_PIT_SPEAKER_DUMMY;
+        self.vm_fd
+            .create_pit2(pit_config)
+            .map_err(Error::KvmIoctl)?;
+
+        // Create the serial console.
+        let interrupt_evt = EventFd::new(libc::EFD_NONBLOCK).map_err(Error::IO)?;
+        let serial = Arc::new(Mutex::new(SerialWrapper(Serial::new(
+            interrupt_evt.try_clone().map_err(Error::IO)?,
+            stdout(),
+        ))));
+        // Register its interrupt fd with KVM.
+        self.vm_fd
+            .register_irqfd(&interrupt_evt, 4)
+            .map_err(Error::KvmIoctl)?;
+
+        // Configure event management for the serial console's events.
+        self.device_mgr.serial_ev_mgr.add_subscriber(serial.clone());
+        self.device_mgr.serial = Some(serial);
+
+        Ok(())
+    }
+
+    /// Configure guest vCPUs.
+    ///
+    /// # Arguments
+    ///
+    /// * `vcpu_cfg` - [`VcpuConfig`](struct.VcpuConfig.html) struct containing vCPU configurations.
+    pub fn configure_vcpus(
+        &mut self,
+        vcpu_cfg: VcpuConfig,
+        kernel_load: KernelLoaderResult,
+    ) -> Result<()> {
+        mptable::setup_mptable(&self.guest_memory, vcpu_cfg.num_vcpus)
+            .map_err(|e| Error::Vcpu(vcpu::Error::Mptable(e)))?;
+
+        let base_cpuid = self
+            .kvm
+            .get_supported_cpuid(KVM_MAX_CPUID_ENTRIES)
+            .map_err(Error::KvmIoctl)?;
+
+        for index in 0..vcpu_cfg.num_vcpus {
+            let vcpu = Vcpu::new(
+                &self.vm_fd,
+                index,
+                self.device_mgr
+                    .serial
+                    .clone()
+                    .expect("Remove this `expect` when the serial console becomes optional"),
+            )
+            .map_err(Error::Vcpu)?;
+
+            // Set CPUID.
+            let mut vcpu_cpuid = base_cpuid.clone();
+            cpuid::filter_cpuid(
+                &self.kvm,
+                index as usize,
+                vcpu_cfg.num_vcpus as usize,
+                &mut vcpu_cpuid,
+            );
+            vcpu.configure_cpuid(&vcpu_cpuid).map_err(Error::Vcpu)?;
+
+            // Configure MSRs (model specific registers).
+            vcpu.configure_msrs().map_err(Error::Vcpu)?;
+
+            // Configure regs, sregs and fpu.
+            vcpu.configure_regs(kernel_load.kernel_load)
+                .map_err(Error::Vcpu)?;
+            vcpu.configure_sregs(&self.guest_memory)
+                .map_err(Error::Vcpu)?;
+            vcpu.configure_fpu().map_err(Error::Vcpu)?;
+
+            // Configure LAPICs.
+            vcpu.configure_lapic().map_err(Error::Vcpu)?;
```

Comment:
I would move the vcpu configuration to a separate function. Might make things more readable.